src/evalfunc.c

```diff
@@ -5452,6 +5455,181 @@ f_getpos(typval_T *argvars, typval_T *rettv)
     getpos_both(argvars, rettv, FALSE, FALSE);
 }
 
+/*
+ * Convert from block_def to string
+ */
+   static char_u *
+block_def2str(struct block_def *bd)
+{
+    char_u *p, *ret;
+    size_t size = bd->startspaces + bd->endspaces + bd->textlen;
+
+    ret = alloc(size + 1);
+    if (ret != NULL)
+    {
+	p = ret;
+	vim_memset(p, ' ', bd->startspaces);
+	p += bd->startspaces;
+	mch_memmove(p, bd->textstart, bd->textlen);
+	p += bd->textlen;
+	vim_memset(p, ' ', bd->endspaces);
+	*(p + bd->endspaces) = NUL;
+    }
+    return ret;
+}
+
+/*
+ * "getregion()" function
+ */
+    static void
+f_getregion(typval_T *argvars, typval_T *rettv)
+{
+    linenr_T		lnum;
+    oparg_T		oap;
+    struct block_def	bd;
+    char_u		*akt;
+    int			inclusive = TRUE;
+    int			fnum = -1;
+    pos_T		p1, p2;
+    pos_T		*fp = NULL;
+    char_u		*str1, *str2;
+    int			is_visual = FALSE;
+
+    if (rettv_list_alloc(rettv) == FAIL)
+	return;
+
+    virtual_op = virtual_active();
+
+    if (check_for_string_arg(argvars, 0) == FAIL
+	    || check_for_string_arg(argvars, 0) == FAIL)
+	return;
+
+    // NOTE: var2fpos() returns static pointer.
+    fp = var2fpos(&argvars[0], TRUE, &fnum, FALSE);
+    if (fp == NULL)
+	return;
+    p1 = *fp;
+
+    fp = var2fpos(&argvars[1], TRUE, &fnum, FALSE);
+    if (fp == NULL)
+	return;
+    p2 = *fp;
+
+    str1 = tv_get_string(&argvars[0]);
+    str2 = tv_get_string(&argvars[1]);
+
+    if (!LT_POS(p1, p2))
+    {
+	// swap position
+	pos_T p;
+
+	p = p1;
+	p1 = p2;
+	p2 = p;
+    }
+
+    is_visual = (str1[0] == 'v' && str1[1] == NUL)
+	|| (str2[0] == 'v' && str2[1] == NUL);
+    if (is_visual)
+    {
+	if (VIsual_mode == 'v')
+	{
+	    // handle 'selection' == "exclusive"
+	    if (*p_sel == 'e' && !EQUAL_POS(p1, p2))
```

Comment:
Can you add some tests for using exclusive `selection`?  There is quite a bit of logic here for handling exclusive selections which are not yet tested.

Reply:
Well, can you add the tests?

Reply:
I am very busy to fix for other comments.

Reply:
You can add the following test:

```
  " exclusive selection
  new
  set selection=exclusive
  call setline(1, ["a\tc", "x\tz", '', ''])
  call cursor(1, 1)
  call feedkeys("\<Esc>v2l", 'xt')
  call assert_equal(["a\t"], getregion('v', '.'))
  call cursor(1, 1)
  call feedkeys("\<Esc>v$G", 'xt')
  call assert_equal(["a\tc", "x\tz", ''], getregion('v', '.'))
  call cursor(1, 1)
  call feedkeys("\<Esc>v$j", 'xt')
  call assert_equal(["a\tc", "x\tz"], getregion('v', '.'))
  call cursor(1, 1)
  call feedkeys("\<Esc>\<C-v>$j", 'xt')
  call assert_equal(["a\tc", "x\tz"], getregion('v', '.'))
  call cursor(1, 1)
  call feedkeys("\<Esc>\<C-v>$G", 'xt')
  call assert_equal(["a", "x", '', ''], getregion('v', '.'))
  set virtualedit=all
  call cursor(1, 1)
  call feedkeys("\<Esc>2lv2lj", 'xt')
  call assert_equal(['      c', 'x   '], getregion('v', '.'))
  set virtualedit&
  set selection&
  bw!
```

Reply:
Thank you!